### ISSUE TITLE
azurerm_storage_account_sas returns invalid query string

### ISSUE THREAD

**Issue by seppi91** [terraform-providers/terraform-provider-azurerm#3764]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform (and AzureRM Provider) Version
terraform v0.11.13
azurerm 1.31.0

### Affected Resource(s)

* `data.azurerm_storage_account_sas`

### Terraform Configuration Files

code from https://medium.com/@bappertdennis/deploy-azure-functions-with-terraform-83ab88c8373c is used

### Expected Behavior

A valid sas_query is generated.

### Actual Behavior

sas_query is evaluated as not valid.

### Steps to Reproduce

create the resources defined in the blog linked above. 

Try to access the url defined by this pattern:
"https://${data.azurerm_storage_account.this.name}.blob.core.windows.net/${azurerm_storage_container.storage_container.name}/${azurerm_storage_blob.storage_blob.name}${data.azurerm_storage_account_sas.storage_sas.sas}"

Azure returns:
> Server failed to authenticate the request. Make sure the value of Authorization header is formed correctly including the signature. RequestId:6d782440-e01e-003d-4114-30197c000000 Time:2019-07-01T13:54:28.2790055Z

1. `terraform apply`

### References

* #0000

**Comment by tombuildsstuff** [comment]
hey @seppi91 

Thanks for opening this issue.

As mentioned in https://github.com/hashicorp/go-azure-helpers/issues/30 at this time the Azure Provider only supports generating SAS Tokens for an Account, rather than for a Service ([which is being tracked in this issue](https://github.com/terraform-providers/terraform-provider-azurerm/issues/1868)). Rather than having multiple issues open tracking the same thing I'm going to close this issue in favour of that one - would you mind subscribing to that one for updates?

Thanks!

**Comment by seppi91** [comment]
Hi @tombuildsstuff and sorry for raising two issues for this thing.

I totally agree to your comment in , https://github.com/hashicorp/go-azure-helpers/issues/30 and closing this issue. But this issue doesn't refer to a service sas token but to an account token which is evaluated as invalid when trying to access the resource.

This is one of the links terraform outputs:
https://ff0cstartereuwdevfmstg.blob.core.windows.net/function-releases/simple.zip?sv=2017-07-29&ss=b&srt=o&sp=r&se=2020–03–21&st=2019–03–21&spr=https&sig=<SECRET>

This is one URL generated by Azure Console:
https://ff0cstartereuwdevfmstg.blob.core.windows.net/function-releases/simple.zip?sp=r&st=2019-07-02T14:39:31Z&se=2019-07-02T22:39:31Z&spr=https&sv=2018-03-28&sig=<SECRET>&sr=b

Calling the URI generated by terraform leads to an error page, where the Azure Console URI works flawless.

What I've noted when looking at these two URIs is that the Azure one doesn't use the "&srt=" parameter even if it's set as "required" in the docs https://docs.microsoft.com/en-us/rest/api/storageservices/constructing-an-account-sas

Do I miss something here or is this a not communicated API change by microsoft?

**Comment by jhosteny** [comment]
I am seeing the same thing. Interestingly, if I take the same parameters being passed into azurerm_storage_account_sas (I logged them) and add them to the test in the go-azure-helpers repo, and run a test case, it produces a token that works for download.

I also took the parameters from the passing test case, and added them explicitly to the provider invocation of `storage.ComputeAccountSASToken` and got a different result. I haven't been able to see why yet.

**Comment by jhosteny** [comment]
@seppi91 I thought I was seeing the same thing, and also used that article as a reference. Did you by any chance copy the code from that article into your .tf file? If so, check the `start` and `expiry` strings. Those are not ascii hyphens, and this is what bit me. It took me an unfortunate amount of time to track that down.

When I fixed that, it worked for me. Note that I got the same thing from the provider as through the portal. I ran a modified local provider with `sv` set to `2017-07-29` to ensure that the signature was exactly the same.
